qtfred/src/ui/dialogs/ShipEditor/ShipPathsDialog.cpp

```diff
@@ -0,0 +1,74 @@
+#include "ShipPathsDialog.h"
+#include "ui_ShipPathsDialog.h"
+#include <ui/util/SignalBlockers.h>
+
+#include <QCloseEvent>
+namespace fso {
+namespace fred {
+namespace dialogs {
+ShipPathsDialog::ShipPathsDialog(QWidget* parent,
+	EditorViewport* viewport,
+	const int ship,
+	const int target_class,
+	const bool departure)
+	: QDialog(parent), ui(new Ui::ShipPathsDialog()),
+	  _model(new ShipPathsDialogModel(this, viewport, ship, target_class, departure)), _viewport(viewport)
+{
+	ui->setupUi(this);
+	connect(ui->pathList, &QListWidget::itemChanged, this, &ShipPathsDialog::changed);
+	connect(this, &QDialog::accepted, _model.get(), &ShipPathsDialogModel::apply);
+	connect(this, &QDialog::rejected, _model.get(), &ShipPathsDialogModel::reject);
+	if (departure) {
+		ui->instructLabel->setText("Restrict departure paths to the following:");
+	} else {
+		ui->instructLabel->setText("Restrict arrival paths to the following:");
+	}
+	updateUI();
+
+	// Resize the dialog to the minimum size
+	resize(QDialog::sizeHint());
+}
+ShipPathsDialog::~ShipPathsDialog() = default;
+void ShipPathsDialog::closeEvent(QCloseEvent* event) {
+	if (_model->query_modified()) {
+		auto button = _viewport->dialogProvider->showButtonDialog(DialogType::Question,
+			"Changes detected",
+			"Do you want to keep your changes?",
+			{DialogButton::Yes, DialogButton::No, DialogButton::Cancel});
+
+		if (button == DialogButton::Cancel) {
+			event->ignore();
+			return;
+		}
+
+		if (button == DialogButton::Yes) {
+			accept();
+			return;
+		}
```

Comment:
Maybe we could do a pair of OK/Cancel buttons in the dialog and automatically accept if they hit OK?  As little friction as possible is better, imho.

Reply:
It does see line 18-19 connect to the ok/cancel buttons on the dialog. This code runs only when the user presses the OS close button

Reply:
Ah, you're right.  Sorry for missing that.  We can still add a confirmation to the cancel button by using this connect statement instead of the default
`connect(ui->OkCancelButtons, &QDialogButtonBox::rejected, this, &yourDialog::reject);`

One of the goals will be to add confirmations to the cancel buttons that don't have them when we get to the polishing stage, anyway.

Reply:
Credit where credits due, I didn't come up with it. I copied it from whoever did the mission specs dialog.